src/openrct2/world/tile_inspector.c

```diff
@@ -0,0 +1,939 @@
+sint32 tile_inspector_track_base_height_offset(sint32 x, sint32 y, sint32 elementIndex, sint8 offset, sint32 flags)
+{
+	rct_map_element *const trackElement = map_get_nth_element_at(x, y, elementIndex);
+
+	if (offset == 0)
+	{
+		return MONEY32_UNDEFINED;
+	}
+
+	if (!trackElement || map_element_get_type(trackElement) != MAP_ELEMENT_TYPE_TRACK)
+	{
+		return MONEY32_UNDEFINED;
+	}
+
+	if (flags & GAME_COMMAND_FLAG_APPLY)
+	{
+		uint8 type = trackElement->properties.track.type;
+		sint16 originX = x << 5;
+		sint16 originY = y << 5;
+		sint16 originZ = trackElement->base_height * 8;
+		uint8 rotation = map_element_get_direction(trackElement);
+		uint8 rideIndex = trackElement->properties.track.ride_index;
+		rct_ride* ride = get_ride(rideIndex);
+		const rct_preview_track* trackBlock = get_track_def_from_ride(ride, type);
+		trackBlock += trackElement->properties.track.sequence & 0x0F;
+
+		uint8 originDirection = map_element_get_direction(trackElement);
+		switch (originDirection)
+		{
+		case 0:
+			originX -= trackBlock->x;
+			originY -= trackBlock->y;
+			break;
+		case 1:
+			originX -= trackBlock->y;
+			originY += trackBlock->x;
+			break;
+		case 2:
+			originX += trackBlock->x;
+			originY += trackBlock->y;
+			break;
+		case 3:
+			originX += trackBlock->y;
+			originY -= trackBlock->x;
+			break;
+		}
+
+		originZ -= trackBlock->z;
+
+		trackBlock = get_track_def_from_ride(ride, type);
+		for (; trackBlock->index != 255; trackBlock++)
+		{
+			sint16 elemX = originX, elemY = originY, elemZ = originZ;
+
+			switch (originDirection)
+			{
+			case 0:
+				elemX += trackBlock->x;
+				elemY += trackBlock->y;
+				break;
+			case 1:
+				elemX += trackBlock->y;
+				elemY -= trackBlock->x;
+				break;
+			case 2:
+				elemX -= trackBlock->x;
+				elemY -= trackBlock->y;
+				break;
+			case 3:
+				elemX -= trackBlock->y;
+				elemY += trackBlock->x;
+				break;
+			}
+
+			elemZ += trackBlock->z;
+
+			map_invalidate_tile_full(elemX, elemY);
+
+			bool found = false;
+			rct_map_element *mapElement = map_get_first_element_at(elemX >> 5, elemY >> 5);
+			do
+			{
+				if (mapElement->base_height != elemZ / 8)
+					continue;
+
+				if (map_element_get_type(mapElement) != MAP_ELEMENT_TYPE_TRACK)
+					continue;
+
+				if ((mapElement->type & MAP_ELEMENT_DIRECTION_MASK) != rotation)
+					continue;
+
+				if ((mapElement->properties.track.sequence & 0x0F) != trackBlock->index)
+					continue;
+
+				if (mapElement->properties.track.type != type)
+					continue;
+
+				found = true;
+				break;
+			} while (!map_element_is_last_for_tile(mapElement++));
+
+			if (!found)
+			{
+				log_error("Track map element part not found!");
+				return MONEY32_UNDEFINED;
+			}
+
+			// track_remove returns here on failure, not sure when this would ever be hit. Only thing I can think of is for when you decrease the map size.
+			openrct2_assert(map_get_surface_element_at(elemX >> 5, elemY >> 5) != NULL, "No surface at %d,%d", elemX >> 5, elemY >> 5);
```

Comment:
Yes, I remember that one. While iterating over a track piece that exists of multiple tiles, it simply adds the offset to the start position, which _could_ be outside of the map. If entire track elements are removed when reducing the map size, then this should never be hit. Another thing that could cause this to be hit is if someone deletes the surface under a track element.